Add rig support and upgrade colmap to version 3.12

lpanaf opened the pull request.

zalo commented:
When testing on a `panorama_sfm.py` project, I'm getting a crash in the "Rotation Averaging" step on Windows in both the CUDA and non-CUDA builds.

The standard colmap 3.12.0 step works properly:
```U:\sfm\colmap-x64-windows-cuda\bin>colmap mapper --database_path ../../PanoramaRigTest/database.db --output_path ../../PanoramaRigTest/sparse --image_path ../../PanoramaRigTest/images --Mapper.ba_use_gpu 1 --Mapper.ba_refine_sensor_from_rig 0 --Mapper.ba_refine_focal_length 0 --Mapper.ba_refine_extra_params 0```

Whereas this branch's glomap fails:
```U:\sfm\glomap-x64-windows-cuda\bin>glomap mapper --database_path ../../PanoramaRigTest/database.db --output_path ../../PanoramaRigTest/sparse --image_path ../../PanoramaRigTest/images --BundleAdjustment.optimize_intrinsics 0 --BundleAdjustment.optimize_rotations 0```

"Optimize Rotations" was disabled in an attempt to disable the Rotation Averaging step, but it doesn't seem to control the same thing 😅 

Here are the full logs:
```
I20250703 15:00:49.399286 28248 panorama_sfm3.py:run:278] [INFO] Matched features in 13.34 seconds.
[INFO] Running Incremental Mapping...
 Loading Images 712 / 712
 Loading Image Pair 8159 / 8159
I20250703 15:00:50.253301 28160 colmap_converter.cc:429] Pairs read done. 430 / 8159 are invalid
I20250703 15:00:50.255123 28160 global_mapper.cc:84] Loaded database
-------------------------------------
Running preprocessing ...
-------------------------------------
I20250703 15:00:50.256477 28160 view_graph_manipulation.cc:249] Decompose relative pose for 0 pairs
I20250703 15:00:50.256886 28160 view_graph_manipulation.cc:303] Decompose relative pose done. 0 pairs are pure rotation
I20250703 15:00:50.258094 28160 timer.cc:87] Elapsed time: 0.00265 [seconds]
-------------------------------------
Running view graph calibration ...
-------------------------------------
I20250703 15:00:50.258298 28160 view_graph_calibration.cc:16] Start ViewGraphCalibrator
I20250703 15:00:50.270112 28160 view_graph_calibration.cc:147] 0 cameras are rejected in view graph calibration
I20250703 15:00:50.270989 28160 view_graph_calibration.cc:181] invalid / total number of two view geometry: 1171 / 3203
-------------------------------------
Running relative pose estimation ...
-------------------------------------
I20250703 15:00:50.272526 28160 image_undistorter.cc:20] Undistorting images..
I20250703 15:00:50.275849 28160 image_undistorter.cc:44] Image undistortion done
I20250703 15:00:50.277224 28160 relpose_estimation.cc:26] Estimating relative pose for 6558 pairs
 Estimating relative pose: 100%
I20250703 15:00:56.503152 28160 relpose_estimation.cc:122] Estimating relative pose done
I20250703 15:00:56.519044 28160 relpose_filter.cc:46] Filtered 3112 relative poses with inlier number < 30
I20250703 15:00:56.519289 28160 relpose_filter.cc:63] Filtered 0 relative poses with inlier ratio < 0.25
I20250703 15:00:56.520732 28160 timer.cc:87] Elapsed time: 6.24854 [seconds]
-------------------------------------
Running rotation averaging ...
-------------------------------------
Traceback (most recent call last):
  File "u:\sfm\panorama_sfm3.py", line 326, in <module>
    run(parser.parse_args())
  File "u:\sfm\panorama_sfm3.py", line 308, in run
    subprocess.run(f"{args.path_to_glomap_cli} mapper --database_path {database_path} --output_path {rec_path} --image_path {image_dir} --BundleAdjustment.optimize_intrinsics 0 --BundleAdjustment.optimize_rotations 0", shell=True, check=True)
  File "U:\anaconda3\Lib\subprocess.py", line 571, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command 'glomap-x64-windows-nocuda\bin\glomap mapper --database_path output7\database.db --output_path output7\sparse --image_path output7\images --BundleAdjustment.optimize_intrinsics 0 --BundleAdjustment.optimize_rotations 0' returned non-zero exit status 3221226505.
````
Here's the repro project if that's useful:
https://drive.google.com/file/d/1Sq3wQf8nXVhVRL4tNjz1SiKUNb4gVsRD/view?usp=sharing

ahojnnes commented on glomap/controllers/global_mapper.cc:
```diff
@@ -215,8 +230,10 @@ bool GlobalMapper::Solve(const colmap::Database& database,
       if (ite != options_.num_iteration_bundle_adjustment - 1)
         run_timer.PrintSeconds();
 
+      // TODO: determine the logic for reconstruction normalization
```
What's the missing normalization logic that needs to be done?

ahojnnes commented:
Thanks for the feedback. I still need to find time to review this. We'll merge it soon and release a new version.

zalo commented:
Ooooh, when SuiteSparse is installed, then I get this on my test dataset on Ubuntu 24.04:
```
I20250714 22:48:02.408242  1671 global_positioning.cc:79] Solving the global positioner problem
W20250714 22:48:06.185361  1671 levenberg_marquardt_strategy.cc:123] Linear solver failure. Failed to compute a step: CHOLMOD warning: Matrix not positive definite.
W20250714 22:48:07.922480  1671 levenberg_marquardt_strategy.cc:123] Linear solver failure. Failed to compute a step: CHOLMOD warning: Matrix not positive definite.
W20250714 22:48:09.453580  1671 levenberg_marquardt_strategy.cc:123] Linear solver failure. Failed to compute a step: CHOLMOD warning: Matrix not positive definite.
W20250714 22:48:10.916043  1671 levenberg_marquardt_strategy.cc:123] Linear solver failure. Failed to compute a step: CHOLMOD warning: Matrix not positive definite.
W20250714 22:48:12.534420  1671 levenberg_marquardt_strategy.cc:123] Linear solver failure. Failed to compute a step: CHOLMOD warning: Matrix not positive definite.
E20250714 22:48:12.534641  1671 trust_region_minimizer.cc:97] Terminating: Number of consecutive invalid steps more than Solver::Options::max_num_consecutive_invalid_steps: 5
I20250714 22:48:12.661639  1671 global_positioning.cc:88] Ceres Solver Report: Iterations: 5, Initial cost: 9.429243e+05, Final cost: 9.429243e+05, Termination: FAILURE
```

Took me forever to track it down to being SuiteSparse 😅 
Seems like there is some aspect of the `levenberg_marquardt_strategy` that is incompatible with SuiteSparse's CHOLMOD in this PR 💀

xiemeilong commented:
@zalo Did you compile colmap with -DBLA_VENDOR=Intel10_64lp? I found it works without -DBLA_VENDOR=Intel10_64lp and installing dependences with --no-install-recommends --no-install-suggests.

ahojnnes commented on glomap/scene/camera_rig.h:
```diff
@@ -0,0 +1,33 @@
+
+namespace glomap {
+
+// struct CameraRig : public colmap::CameraRig {
```
Remove this file?

ahojnnes commented on glomap/io/pose_io.cc:
```diff
@@ -171,7 +188,7 @@ void WriteGlobalRotation(const std::string& file_path,
     file << image.file_name;
     for (int i = 0; i < 4; i++) {
-      file << " " << image.cam_from_world.rotation.coeffs()[(i + 3) % 4];
+      file << " " << image.CamFromWorld().rotation.coeffs()[(i + 3) % 4];
```
nit: store as temporary, otherwise the pose is recomposed 4 times.

ahojnnes commented on glomap/io/pose_io.cc:
```diff
@@ -118,6 +130,7 @@ void ReadRelWeight(const std::string& file_path,
   LOG(INFO) << counter << " weights are used are loaded" << std::endl;
 }
 
+// TODO: now, it does not care about the frames
```
I don't understand this TODO. Either resolve or provide more context. Thanks.

ahojnnes commented on glomap/controllers/rotation_averager.cc:
```diff
@@ -1,11 +1,16 @@
                             std::unordered_map<image_t, Image>& images,
                             const RotationAveragerOptions& options) {
-  view_graph.KeepLargestConnectedComponents(images);
+  view_graph.KeepLargestConnectedComponents(frames, images);
```
Can you remove the old function?

lpanaf replied:
Sounds good. I refactored the structure of the Image / Frame class so that is_registered is now a property of Frame instead of Image

ahojnnes commented on glomap/controllers/rotation_averager.cc:
```diff
@@ -46,16 +51,152 @@ bool SolveRotationAveraging(ViewGraph& view_graph,
+  // By default, run trivial rotation averaging for rigged cameras if some
+  // cam_from_rig are not estimated Check if there are rigs with non-trivial
+  // cam_from_rig
+  // bool run_trivial_ra = false;
```
Remove if not needed.

lpanaf commented:
@ahojnnes Thanks for the feedback. I have incorporated the changes into the new commits. I also updated the implementation of image pruning based on rig structures.

ahojnnes commented:
@lpanaf Thank you very much again for this change. I finally had the time to test this on some datasets. It seems to work well. I'll push to this branch to update to the very latest colmap 3.12.5 version.

ahojnnes commented on glomap/controllers/global_mapper.cc:
```diff
@@ -14,11 +15,28 @@
 
 namespace glomap {
 
+// TODO: Rig normalizaiton has not be done
```
@lpanaf Not sure I understand this comment. Can you address it before merging?

lpanaf replied:
Before, I intended not to normalize the reconstruction if the rig calibration is provided. However, after the discussion last time, I think it would be nice to keep it consistent with COLMAP: I should always normalize the scale.

ahojnnes commented on glomap/controllers/global_mapper.cc:
```diff
@@ -46,6 +64,7 @@ bool GlobalMapper::Solve(const colmap::Database& database,
   }
 
   // 2. Run relative pose estimation
+  //   TODO: Use generalized relative pose estimation for rigs.
```
COLMAP now has a EstimateGeneralizeRelativePose function for this purpose.

lpanaf replied:
Thanks for the suggestion! Maybe I will leave this TODO for another pull request. The problem I am thinking of is that the estimator is only for the case where two rigs have the same scale.  I would need to distinguish between these cases from the case with rigs with arbitrary scales.

ahojnnes commented on glomap/estimators/cost_function.h:
```diff
@@ -40,6 +40,106 @@ struct BATAPairwiseDirectionError {
+
+  // TODO: add covariance
+  const Eigen::Vector3d translation_obs_;
+  const Eigen::Quaterniond& rig_from_world_rot_;  // = c_R_w^T * c_t_r
```
```suggestion
  const Eigen::Quaterniond rig_from_world_rot_;  // = c_R_w^T * c_t_r
```

ahojnnes commented on glomap/estimators/gravity_refinement.cc:
```diff
@@ -95,67 +120,65 @@ void GravityRefiner::RefineGravity(const ViewGraph& view_graph,
 
   // image_id: (mistake, total)
-  std::unordered_map<image_t, std::pair<int, int>> image_counter;
+  std::unordered_map<frame_t, std::pair<int, int>> frame_counter;
```
```suggestion
  std::unordered_map<frame_t, std::pair<int, int>> frame_counter;
  frame_counter.reserve(frames.size());
```

AlbertoJaenal commented:
Tested the code  (commit [79af753](https://github.com/colmap/glomap/pull/201/commits/79af75355871560562b8e2a4e81d4dbdf61178b8)) on a couple of public datasets and on my own data and it works well. Thanks!
Any prognostics on implementing something similar to colmap's `Mapper.ba_refine_sensor_from_rig`? Have not really seen through the code, so don´t know how feasible it is.

GibsDev commented:
I would like to give this a try soon. Can it be used with colmap 3.12.5, or should I be using the latest main branch?

lpanaf commented:
Hi @GibsDev, I originally developed this pull request based on git id `f2137129796fe3b0edfbb379220b375bcf8a635e`. It should very likely be fine with 3.12.5.

ahojnnes commented:
I already updated this branch against 3.12.5. It should work.

lpanaf commented:
Hi @DuVogel87, you would need to write the imported rig info into the COLMAP database, then run `glomap mapper`. `--BundleAdjustment.optimize_rig_poses=1` is not a necessary flag if rig calibrations are of good quality.

GibsDev commented:
https://github.com/colmap/glomap/pull/201#issuecomment-3322164817

I discovered that you need to make sure to set the translation to NaN for `--BundleAdjustment.optimize_rig_poses` to work. Earlier in my script I am setting the approximate rotation and translation for the rig. I was setting 0's for the translation, but it needs to be NaNs instead.

The zeroed out translation worked with colmap, so it might be good to keep that behavior with glomap to keep it consistent and prevent other people from running into this problem.

lpanaf commented:
Hi @GibsDev, thanks for sharing your experience. Do you mind sharing some data so that we can play around with it? If you set the translation to be NaN, glomap will try to estimate the translation from scratch, so you can observe the difference. However, we expect it to optimize together with other properties in the bundle adjustment part, which should behave similarly to colmap.

GibsDev commented:
I'm afraid I don't have a dataset I can share at the moment without special permission.

In further testing discovered that the rig estimation using glomap was not as effective/accurate as colmap. In one of my test datasets, I was getting 0.84px mean reprojection error with colmap and upwards of 2.3px mean reprojection error with glomap when allowing them to calculate the rig poses from "scratch".

However, when I manually fed the rig poses into the rig config, glomap actually managed to beat colmap for mean reprojection error at 0.75px!

GibsDev commented:
Results have been fantastic in my testing so far! Being able to seed the rig rotation and translation with a very good estimate and then allowing it to refine it seems to be working really well!

The changes to the pruning have also been beneficial. I have one dataset that previously was failing about 80% of the time  due to all of the tracks being filtered out (without changing any settings between runs). Maybe I'm just getting lucky, but it seems to be much more stable!

Well done on this!
